htlcswitch/hop/payload.go

```diff
@@ -120,21 +120,20 @@ func NewPayloadFromReader(r io.Reader) (*Payload, error) {
 	if err != nil {
 		// Promote any required type failures into ErrInvalidPayload.
 		if e, required := err.(tlv.ErrUnknownRequiredType); required {
-			// NOTE: Sigh. If the sender included a next hop whose
-			// value is zero, this would be considered invalid by
-			// our validation rules below. It's not totally clear
-			// whether this required failure should take precedence
-			// over the constraints applied by known types.
-			// Unfortunately this is an artifact of the layering
-			// violation in placing the even/odd rule in the parsing
-			// logic and not at a higher level of validation like
-			// the other presence/omission checks.
-			//
-			// As a result, this may need to be revisted if it is
-			// decided that the checks below overrule an unknown
-			// required type failure, in which case an
-			// IncludedViolation should be returned instead of the
-			// RequiredViolation.
+			// If the parser returned an unknown required type
+			// failure, we'll first check that the payload is
+			// properly formed according to our known set of
+			// contraints. If an error is discovered, this overrides
+			// thre required type failure.
+			nextHop := lnwire.NewShortChanIDFromInt(cid)
```

Comment:
nit: maybe move the `nextHop` line to before `DecodeWithParsedTypes` to reduce the amount of duplicated lines?

Reply:
the assignment has to be done after `DecodeWithParsedTypes` otherwise `cid` will always be zero since it hasn't been parsed yet. thought about doing it between `DecodeWithParsedTypes` and the error check but wasn't happy with the aesthetics :P